tools/curl/gurl/curlparse/__init__.py

```diff
@@ -1,18 +1,99 @@
+from dataclasses import dataclass
+import re
 
+import gurl.hexdealer as he
 
-def curl_trace_block_iterator(curl_content):
+"""
+A block is a send or receive item
+A block represent bytes, so bytearray must be intended
+A block can has meta data associated
+"""
+@dataclass
+class Block:
+    kind: str
+    item: str
+    bytes_size: int
+    data: bytes
+    meta:list=None
+
+
+def _extract_item(line):
+    regex = r"(=>|<=)\s+(Send|Recv)\s+([\w\s]+),\s+(\d+)"
+    match = re.search(regex, line)
+    kind = match.groups()[1]
+    item = match.groups()[2]
+    size = int(match.groups()[3])
+    return kind, item, size
+
+
+def curl_trace_block_iterator(curl_content:bytes):
     if not curl_content or curl_content == '':
         yield from ()
         return
-    content = []
-    yil = False
+    kind = None
+    item = None
+    size = 0
+    content = bytearray()
+    meta = []
+    last_info = False
     for line in curl_content.strip().split(b'\n'):
-        if line.startswith(b'=') or line.startswith(b'<'): # any block
+        if line.startswith(b'=='):
+            last_info = True
+            clean = line.replace(b"== ", b"")
+            meta.append(clean.decode("utf-8"))
+        elif line.startswith(b'=>') or line.startswith(b'<='):
+            last_info = False
             if content:
-                yield b'\n'.join(content)
-                content = []
-            content.append(line)
+                if meta:
+                    yield Block(kind, item, size, bytes(content), meta)
+                else:
+                    yield Block(kind, item, size, bytes(content))
+                content = bytearray()
+                meta = []
+            kind, item, size = _extract_item(line.decode("utf-8"))
+        elif last_info: # append to info
+            meta[-1] = meta[-1] + "\n" + line.decode("utf-8")
         else: # append to block
-            content.append(line)
-    if content:
-        yield b'\n'.join(content)
\ No newline at end of file
+            hex_part = he.extract_hex_from_curl(line)
+            b = bytes.fromhex(hex_part.decode("utf-8"))
+            content.extend(b)
+    if content and meta:
+        yield Block(kind, item, size, bytes(content), meta)
+    elif content:
+        yield Block(kind, item, size, bytes(content))
+
+
+@dataclass
+class MetaReqRes:
+    meta: list
+    req: list
+    res: list
+    is_ssl: bool = False
+
+
+def curl_trace_reqres_iterator(block_iterator):
```

Comment:
If you need to make sure your readers understand is an iterator is better to use a type annotation in this case and change the name to `blocks`.  I would be fine without the type annotation anyway.

Reply:
True, Changing